File: go/src/leetcode/7.reverse-integer.go
/*
 * @lc app=leetcode id=7 lang=golang
 *
 * [7] Reverse Integer
 */
package main

func reverse(x int) int {
	// 1111 ...... 1111
	// 2147483648  			uint32
	// 0111 ...... 1111		int32
	// 2147483648 - 1 = 2147483647	int32Max
	const int32Max = int32(^uint32(0) >> 1)

	// 1111 ...... 1111
	// -(2147483648 - 1)  int32
	// 1000 ...... 0000
	// -2147483648        int32		int32Min
	const int32Min = ^int32Max

	var pop int32
	var reverse int32
	pop, reverse = 0, 0

	if x == 0 {
		return x
	}

	for x != 0 {
		pop = int32(x % 10)
		x /= 10

		if reverse > int32Max/10 || (reverse == int32Max/10 && pop > 7) {
			return 0
		}
		if reverse < int32Min/10 || (reverse == int32Max/10 && pop < -8) {
			return 0
		}
		reverse = reverse*10 + pop

	}

	// if x < 0 {
	// 	return -1 * x
	// }
	return int(reverse)
}

// func main() {
// 	print(reverse(-12))
// }

File: python/7.reverse-integer.py
#
# @lc app=leetcode id=7 lang=python3
#
# [7] Reverse Integer
#

import sys


class Solution:
    def reverse(self, x: int) -> int:
        # intMaxSYS = sys.maxsize
        # print(intMaxSYS)

        intMax32 = int((1 << 32)/2) - 1
        # print(type(intMax32))
        # print(intMax32)

        intMin32 = ~intMax32
        # print(type(intMin32))
        # print(intMin32)

        reverse = 0
        pop = 0

        if x == 0:
            return 0

        if x > 0:
            while x != 0:
                pop = x % 10
                x //= 10

                if reverse > intMax32//10 or (reverse == intMax32 and pop > 7):
                    return 0

                reverse = reverse * 10 + pop

            return reverse

        if x < 0:
            while x != 0:
                pop = x % -10
                x = -(x // -10)

                if reverse < -(intMin32//-10) or (reverse == intMin32 and pop < -8):
                    return 0

                reverse = reverse * 10 + pop
            return reverse


if __name__ == "__main__":
    # Solution().reverse(-123)
    Solution().reverse(-2147583647)
    # Solution().reverse(2147583647)
